### ISSUE TITLE
feature request: wolfSSL support

### ISSUE THREAD

**Issue by benny-de** [acmesh-official/acme.sh#4537]
On embedded devices like OpenWRT wifi routers is often not much free space e.g. 1,2 MB.
OpenWRT switched to wolfSSL. Yes, you can install openssl too or instead of wolfssl on openwrt devices, but with 1,2 MB free space it is a problem.
So it would be nice if acme.sh could support wolfSSL too.
Thank you very much.

**Comment by Neilpang** [comment]
does the wolfssl have a cli ?  just like the `openssl` command.

**Comment by benny-de** [comment]
openwrt installs these wolfssl packages:

"# opkg list-installed |grep wolfssl"
libustream-wolfssl20201210 - 2022-12-08-9217ab46-2
libwolfssl5.5.4.ee39414e - 5.5.4-stable-1
px5g-wolfssl - 6.2
wpad-basic-wolfssl - 2022-01-16-cff80b4f-15.2

=> "# px5g"
PX5G X.509 Certificate Generator Utilit using WolfSSL
Usage: [eckey|rsakey|selfsigned]

https://openwrt.org/packages/pkgdata/px5g-wolfssl
https://github.com/openwrt/openwrt/blob/openwrt-21.02/package/utils/px5g-wolfssl/px5g-wolfssl.c

maybe this could help?

there is https://github.com/wolfSSL/wolfCLU#readme
but this is not part of openwrt default image.

**Comment by prologic** [comment]
There is a CLI for wolfSSL: https://github.com/wolfSSL/wolfCLU#readme

**Comment by prologic** [comment]
It is involved as `wolfssl`

**Comment by prologic** [comment]
Is anyone interested in this? I've successfully built a package for [uLinux](https://git.mills.io/prologic/ulinix) for the `wolfssl` CLI so I'm thinking about seeing if I can hack acme.sh to work with wolfssl.

**Comment by prologic** [comment]
So one thing I've immediately noticed is that wolfssl lacks a `ec` mode, however it does have `ecc`. It also doesn't have a `-text` option. I'm not sure it can display a human-readable version of a private key?

**Comment by Neilpang** [comment]
> Is anyone interested in this? I've successfully built a package for [uLinux](https://git.mills.io/prologic/ulinix) for the `wolfssl` CLI so I'm thinking about seeing if I can hack acme.sh to work with wolfssl.

Yes,  search for the functions that are using `openssl` command inside,  if you can implement them all with wolfssl command, then that should work.

**Comment by neilpang** [comment]
I did a full feasibility analysis of this, including building wolfCLU (v0.2.0 era master) against wolfSSL `--enable-wolfclu` and testing the critical operations against openssl output.

**Feasibility: confirmed.** wolfCLU (the official wolfSSL CLI, invoked as `wolfssl`) now covers almost everything acme.sh needs:

- `x509` has `-text`, `-dates`, `-serial`, `-fingerprint`, `-subject`, `-issuer` (the missing `-text` mentioned above has been added since)
- `req` supports `-new -key -subj -config -extensions`, so CSR generation with SANs works
- `genkey rsa` / `ecparam` replace `genrsa` / EC key generation (different syntax, wrappable)
- `dgst -sign` (RSA and ECC) produces standard signatures -- verified interoperable with openssl
- HMAC-SHA256 for External Account Binding works: `wolfssl dgst -sha256 -hmac -mackey hexkey:<hex>` -- output byte-identical to `openssl dgst -sha256 -mac HMAC -macopt hexkey:<hex>`
- `pkcs12`/`pkcs8`/`s_server` exist

Three real gaps, all solvable with one small POSIX-shell DER/TLV parser (~40 lines, prototype tested on dash, output byte-identical to the current openssl-based extraction):

1. No `asn1parse`: ECDSA JWS signatures need r||s extracted from the DER signature. Parsing `SEQUENCE { INTEGER r, INTEGER s }` in shell works and matches `openssl asn1parse` exactly.
2. No `rsa -text`: the RSA JWK needs publicExponent. Parse it from `pkey -pubout -outform DER` (SPKI) with the same parser; modulus comes from `rsa -modulus` directly.
3. No `ec -text`: the EC JWK needs the x,y point and curve. Same parser on the EC SPKI DER.

The right implementation is an abstraction layer (one `_ssl_*` primitive per operation: keygen, csr, sign, hmac, cert-field, jwk-extract) with openssl and wolfssl backends, not per-call-site branching. Plus a CI image with wolfCLU so the code path stays tested.

**Why I am not implementing it now: distribution.** There is no prebuilt wolfCLU anywhere -- no binaries in its GitHub releases, no Debian/Ubuntu package, and zero hits in the OpenWrt feeds. Worse, wolfCLU requires wolfSSL built with `--enable-wolfclu`, which distro/OpenWrt libwolfssl builds do not enable -- so a user has to cross-compile both wolfSSL and wolfCLU themselves. On the 1.2 MB routers this issue was opened for, that is strictly harder than installing `openssl-util`, and the second wolfSSL copy erases the size benefit.

So the blocker is not acme.sh, it is packaging. Closing for now. If wolfCLU lands in the OpenWrt feed (or ships prebuilt binaries) and there is still demand, we will reopen -- the analysis above is the implementation roadmap.
